src/ophyd_async/core/standard_readable.py

```diff
@@ -1,72 +1,168 @@
-        )
+        return await merge_gathered_dicts([sig.read() for sig in self._readables])
+
+    @property
+    def hints(self) -> Hints:
+        return self._hints
+
+    @contextmanager
+    def add_children_as_readables(
+        self,
+        wrapper: Optional[Type[Union["ConfigSignal", "HintedSignal"]]] = None,
+    ) -> Generator[None, None, None]:
+        dict_copy = self.__dict__.copy()
+
+        yield
+
+        # Set symmetric difference operator gives all newly added items
+        new_attributes = dict_copy.items() ^ self.__dict__.items()
+        new_signals: List[SignalR] = [x[1] for x in new_attributes]
+
+        self._wrap_signals(wrapper, new_signals)
+
+    def add_readables(
+        self,
+        wrapper: Type[Union["ConfigSignal", "HintedSignal"]],
+        *signals: SignalR,
+    ) -> None:
+
+        self._wrap_signals(wrapper, signals)
+
+    def _wrap_signals(
+        self,
+        wrapper: Optional[Type[Union["ConfigSignal", "HintedSignal"]]],
+        signals: Sequence[SignalR],
+    ):
+
+        for signal in signals:
+            obj: Union[SignalR, "ConfigSignal", "HintedSignal"] = signal
+            if wrapper:
+                obj = wrapper(signal)
+
+            if isinstance(obj, AsyncReadable):
+                self._readables.append(obj)
+
+            if isinstance(obj, AsyncConfigurable):
+                self._configurables.append(obj)
+
+            if isinstance(obj, AsyncStageable):
+                self._stageables.append(obj)
+
+            if isinstance(obj, HasHints):
+                new_hint = obj.hints
+
+                # Merge the existing and new hints, based on the type of the value.
+                # This avoids default dict merge behaviour that overrided the values;
+                # we want to combine them when they are Sequences, and ensure they are
+                # identical when string values.
+                for key, value in new_hint.items():
+                    if isinstance(value, Sequence):
+                        if key in self._hints:
+                            self._hints[key] = (  # type: ignore[literal-required]
+                                self._hints[key]  # type: ignore[literal-required]
+                                + value
+                            )
+                        else:
+                            self._hints[key] = value  # type: ignore[literal-required]
+                    elif isinstance(value, str):
+                        if key in self._hints:
+                            assert (
+                                self._hints[key]  # type: ignore[literal-required]
+                                == value
+                            ), "Hints value may not be overridden"
+                        else:
+                            self._hints[key] = value  # type: ignore[literal-required]
+                    else:
+                        raise AssertionError("Unknown type in Hints dictionary")
```

Comment:
Should raise `TypeError` and put the obj.name, key and value in the error message